src/core/trulens/core/experimental/__init__.py

```diff
@@ -0,0 +1,452 @@
+                for frame in self.locked_by:
+                    locked_frames += (
+                        f"  {python_utils.code_line(frame, show_source=True)}\n"
+                    )
+
+                raise ValueError(
+                    f"Feature flag has already been set to {self.value} and cannot be changed. "
+                    f"It has been locked here:\n{locked_frames}"
+                )
+
+            self.value = value
+
+        if lock:
+            self.locked_by.add(python_utils.external_caller_frame(offset=1))
+
+        return self.value
+
+    def get(self, lock: bool = False) -> T:
+        """Get the value of this setting.
+
+        If lock is True, lock the setting.
+        """
+
+        return self.set(lock=lock)
+
+    def lock(self, value: Optional[T] = None) -> T:
+        """Lock the value of this setting.
+
+        If a value is provided, attempt to set the setting first to that value.
+
+        Raises:
+            ValueError: If the setting has already been locked and the value is
+                different from the current value.
+        """
+
+        return self.set(value, lock=True)
+
+
+class _Settings:
+    """A collection of settings to enable/disable experimental features.
+
+    A feature can be enabled/disabled and/or locked so that once it is set, it
+    cannot be changed again. Locking is necessary for some features like OTEL
+    tracing as once components have been instrumented with old or new tracing,
+    the instrumentation cannot be changed.
+    """
+
+    def __init__(self):
+        self.settings: Dict[Feature, _Setting[bool]] = defaultdict(
+            lambda: _Setting(default=False)
+        )
+        """The settings for the experimental features."""
+
+    def is_locked(self, flag: Union[Feature, str]) -> bool:
+        """Determine if the given setting is locked."""
+
+        return self.settings[Feature(flag)].is_locked
+
+    def set(
+        self,
+        flag: Union[Feature, str],
+        *,
+        value: Optional[bool] = None,
+        lock: bool = False,
+    ) -> bool:
+        """Get/Set the given feature flag to the given value.
+
+        Sets the flag to the given value if the value parameter is set. Locks
+        the flag if the lock parameter is set to True.
+
+        Raises:
+            ValueError: If the flag was already locked to a different value.
+        """
+
+        return self.settings[Feature(flag)].set(value=value, lock=lock)
+
+    def lock(
+        self, flag: Union[Feature, str], *, value: Optional[bool] = None
+    ) -> bool:
+        """Lock the given feature flag to the given value.
+
+        If the value is not provided, lock the flag to its current value.
+
+        Raises:
+            ValueError: If the flag has already been locked to a different value.
+        """
+
+        return self.set(flag, value=value, lock=True)
+
+    def get(self, flag: Union[str, Feature], *, lock: bool = False) -> bool:
```

Comment:
Why do we need lock param for get operation?

Reply:
That is actually the more common use case of locking. If you need to configure something differently based on the value of a feature flag and if this configuration is permanent so it cannot be changed, you'd lookup the feature flag while locking it:
```python
if settings.get(Feature.OTEL_TRACING, lock=True):
   # do instrumentation one way
else:
   # do instrumentation in some other way
```